app/src/main/java/com/stevesoltys/seedvault/plugins/StoragePluginManager.kt

```diff
@@ -0,0 +1,162 @@
+/*
+ * SPDX-FileCopyrightText: 2024 The Calyx Institute
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.stevesoltys.seedvault.plugins
+
+import android.content.Context
+import android.net.ConnectivityManager
+import android.net.NetworkCapabilities.NET_CAPABILITY_INTERNET
+import androidx.annotation.WorkerThread
+import com.stevesoltys.seedvault.getStorageContext
+import com.stevesoltys.seedvault.permitDiskReads
+import com.stevesoltys.seedvault.plugins.saf.DocumentsProviderStoragePlugin
+import com.stevesoltys.seedvault.plugins.saf.DocumentsStorage
+import com.stevesoltys.seedvault.plugins.saf.SafFactory
+import com.stevesoltys.seedvault.plugins.webdav.WebDavFactory
+import com.stevesoltys.seedvault.settings.SettingsManager
+import com.stevesoltys.seedvault.settings.StoragePluginEnum
+
+abstract class StorageProperties<T> {
+    abstract val config: T
+    abstract val name: String
+    abstract val isUsb: Boolean
+    abstract val requiresNetwork: Boolean
+
+    @WorkerThread
+    abstract fun isUnavailableUsb(context: Context): Boolean
+
+    /**
+     * Returns true if this is storage that requires network access,
+     * but it isn't available right now.
+     */
+    fun isUnavailableNetwork(context: Context, allowMetered: Boolean): Boolean {
+        return requiresNetwork && !hasUnmeteredInternet(context, allowMetered)
+    }
+
+    private fun hasUnmeteredInternet(context: Context, allowMetered: Boolean): Boolean {
+        val cm = context.getSystemService(ConnectivityManager::class.java) ?: return false
+        val isMetered = cm.isActiveNetworkMetered
+        val capabilities = cm.getNetworkCapabilities(cm.activeNetwork) ?: return false
+        return capabilities.hasCapability(NET_CAPABILITY_INTERNET) &&
+            (allowMetered || !isMetered)
+    }
+}
+
+class StoragePluginManager(
+    private val context: Context,
+    private val settingsManager: SettingsManager,
+    safFactory: SafFactory,
+    webDavFactory: WebDavFactory,
+) {
+
+    private var _appPlugin: StoragePlugin<*>?
```

Comment:
Do we plan on prefixing with `_` for private variables throughout the app? I think we don't do that elsewhere right now do we?